Allow users to copy a form to different app

kaapstorm opened the pull request:
This was written to allow the L10K project to copy forms and their case management from a development app to a live app.

It is hidden behind a feature flag. If it's useful enough, we could consider dropping the flag.

For now I think the flag offsets the risk of preserving current functionality that allows users to copy forms from a module of one case type to a module of a different case type. (See [this mysterious docstring][1] written by @dannyroberts on 2013-12-12.) If you feel that we should change this behaviour, that seems wise, but maybe @dannyroberts can weigh in first.

  [1]: https://github.com/dimagi/commcare-hq/blob/9bf564262910a898b4afa102efa831ce0206a7fb/corehq/apps/app_manager/models.py#L6032

Labeling "Hold off" to give Cody a chance to check this on Staging.

buddy @mkangia cc @snopoke

kaapstorm committed:
Add toggle

kaapstorm committed:
Support src/dst app

kaapstorm committed:
Typo

kaapstorm committed:
Test

kaapstorm committed:
Cleanup

kaapstorm committed:
View

kaapstorm committed:
Test

kaapstorm committed:
ViewModel

kaapstorm committed:
Template

kaapstorm committed:
Stickler

mkangia commented on corehq/apps/app_manager/models.py:
```diff
@@ -3573,8 +3573,8 @@ def add_insert_form(self, from_module, form, index=None, with_source=False):
-                    app = self.get_app()
-                    select_chain = get_select_chain(app, from_module, include_self=False)
+                    from_app = from_module.get_app()
+                    select_chain = get_select_chain(from_app, from_module, include_self=False)
```
nit: i am not sure how much adding `from_` adds value to app here. You could just change to
`app = from_module.get_app()`
Instead add documentation to the method that makes it clear that from_module can be within the same app or a different app as well considering that's not obvious.

mkangia commented on corehq/apps/app_manager/models.py:
```diff
@@ -6025,17 +6025,19 @@ def scrub_source(self, source):
 
         """
+        if to_app is None:
+            to_app = self
```
nit: `to_app = to_app or self`

mkangia commented on corehq/apps/app_manager/views/forms.py:
```diff
@@ -122,23 +122,24 @@ def copy_form(request, domain, app_id, form_unique_id):
     app = get_app(domain, app_id)
     form = app.get_form(form_unique_id)
     module = form.get_module()
+    to_app = get_app(domain, request.POST['to_app_id']) if request.POST.get('to_app_id') else app
```
Do we want to handle in case get_app fails?

kaapstorm replied:
get_app handles if it fails; it raises a 404 if app_id is not found. 

(Which, to be honest, might be a little easy on the user, considering app_id comes from HQ, so if it's not found it's probably because the user has intentionally been messing with the POST values. Unfortunately computers don't offer the option of giving the user a strong negative user experience :smiling_imp: but if they invent keyboards or mice that can taze the user, please let me know. (Also curious why anyone would buy one of those.))

mkangia replied:
will do :)

mkangia commented on corehq/apps/app_manager/views/forms.py:
```diff
@@ -654,6 +669,12 @@ def get_form_view_context_and_template(request, domain, form, langs, messages=me
 
+    if (
+            toggles.COPY_FORM_TO_APP.enabled(domain) or
+            toggles.COPY_FORM_TO_APP.enabled(request.user.username)
```
I think you can use `enabled_for_request` here. `toggles.COPY_FORM_TO_APP.enabled_for_request(request)`

mkangia commented on corehq/apps/app_manager/tests/test_views.py:
```diff
@@ -218,3 +220,33 @@ def test_default_new_app(self, mock):
+        other_app.save()
+        self.addCleanup(lambda: Application.get_db().delete_doc(other_app.id))
+
+        linked_app = create_linked_app(self.domain.name, self.app.id, self.domain.name, 'LinkedApp')
```
I think we have been trying to bring some consistency to using project for `Domain` object and `domain` for its name. If you can squeeze that change here for the tests in this file.

mkangia commented on corehq/apps/app_manager/static/app_manager/js/forms/copy_form_to_app.js:
```diff
@@ -0,0 +1,50 @@
+            var selected = _.filter(self.apps(), function (app) {
+                return app.id === self.selectedAppId();
+            });
+            return selected.length ? selected[0] : undefined;
```
i think you need https://underscorejs.org/#find here.

kaapstorm replied:
`<3`

mkangia reviewed (commented):
couple of things
1. nit: i think we can avoid adding `to_` and `from_` naming change at places. But fine if you want to keep it. 
Also can add documentation to make it clear why we are fetching app from module instead of using self because it can come from different application.

2. In the view this now makes all modules available for all apps in case the flag is enabled. In case a user wants to add module within the same app and has the FF enabled, they would need to traverse this long list even when they dont need to. If it does not complicate things, would it make sense to keep list of modules in the current app available as-is and the list of modules from apps other than current separate for the purpose this PR is addressing?

mkangia commented:
Build failure looks related
```
======================================================================
FAIL: corehq.apps.app_manager.tests.test_views:TestViews.test_get_apps_modules
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/vendor/lib/python2.7/site-packages/mock/mock.py", line 1305, in patched
    return func(*args, **keywargs)
  File "/mnt/commcare-hq-ro/corehq/apps/app_manager/tests/test_views.py", line 244, in test_get_apps_modules
    self.assertEqual(len(apps_modules), 2, 'get_apps_modules should only return normal Applications')
AssertionError: get_apps_modules should only return normal Applications
----------------------------------------------------------------------
```

mkangia commented:
Had a look at the test is failing because the deleted app is getting included because of [this](https://github.com/dimagi/commcare-hq/blob/d5a7b7c77e4b36a819a8e866d804c1e5fb5ce25c/corehq/apps/app_manager/util.py#L70)

i tried
```
def get_apps_modules(domain):
    """
    Returns a domain's apps and their modules
    """
    return [
        {
            'app_id': app.id,
            'name': app.name,
            'modules': [{
                'module_id': module.id,
                'name': clean_trans(module.name, app.langs),
            } for module in app.modules]
        }
        for app in get_apps_in_domain(domain, include_remote=False)
        if app.doc_type == 'Application' and not app.is_deleted()  # No linked apps, no deleted apps
    ]
```
but obviously that failed as well.

kaapstorm commented:
> nit: i think we can avoid adding to_ and from_ naming change at places. But fine if you want to keep it.
> Also can add documentation to make it clear why we are fetching app from module instead of using self because it can come from different application.

There is a little bit of philosophy behind my choice here, and your last sentence there helps explains it. I am a **huge** fan of documentation. It serves many purposes, from [guiding your design][1], to [empowering contributors and community][2]. **But** it also opens a vulnerability to a new kind of tech debt: the possibility of your code and your documentation diverging. And so avoiding documentation by writing blindingly obvious code is a good thing. Considering modules can now come from different apps, and modules are _already_ named `from_module` and `to_module`, it not only seemed consistent to name apps `from_app` and `to_app`, it also seemed wise.

If you think that the line 
```python
    from_app = from_module.get_app()
```
does not make it obvious enough that modules can come from different apps, let me know, and I'll add a comment as well, but I hope that my variable names are enough to explain what I'm doing.


  [1]: https://www.kennethreitz.org/essays/documentation-is-king
  [2]: https://www.youtube.com/watch?v=bQSR1UpUdFQ

kaapstorm committed:
Feedback

kaapstorm committed:
Default app, module to current

kaapstorm committed:
s/domain/project/

kaapstorm commented:
`is_current` is a pass at improving the less-than-ideal experience you pointed out here:

> If it does not complicate things, would it make sense to keep list of modules in the current app
> available as-is and the list of modules from apps other than current separate for the purpose this
> PR is addressing?

The UI now has the current app and current module selected by default in the two select dropdowns. Now coping a form to the same module is just a matter of clicking "Copy", and copying it to a different module in the same app is a matter of selecting the module only, just like when the FF is disabled. 

By defaulting the selection to the current app and module, we have only one module dropdown instead of two, which seems simpler.

I don't think it's exactly what you suggested, but I hope it solves the problem you identified, or is at least an improvement.

kaapstorm commented:
(There is one more commit coming, to ensure that deleted apps aren't shown.)

mkangia commented:
Cool.. if it makes the experience smooth, great!

also i would usually add snapshots, quick video in the description to help understand/review UI changes better. Not sure how useful that is. For ex: https://github.com/dimagi/commcare-hq/pull/20939. Don't need to add it here but just mentioning if you think it would be useful.

kaapstorm commented:
Lol, I seriously considered that, especially for the "is_current" change, but then you'd see how my local environment's styling is a mess, and I haven't got around to fixing it. :rofl: :see_no_evil: :grimacing:

![defaults_to_current_app module_take2](https://user-images.githubusercontent.com/708421/41589014-aafaa53a-73b3-11e8-81c1-0e7183b62563.gif)
^^^ Here it is, fwiw. The video shows the app set to a different app than the current one. After a refresh, the select dropdowns default to the current app, and the current module (whose name is black on navy blue, so not so easy to see).

kaapstorm committed:
Fix test

kaapstorm commented:
OK, I think we're done. (But wait for Travis anyway.) The test was failing because a previous test was creating an app and not cleaning it up.

(Also added a comment to explain why `get_apps_modules()` uses `app.doc_type` instead of `app.get_doc_type()`. Dropped the `include_remote=False` param because remote apps get returned by the DB anyway and then skipped, so we might as well skip all non-Application instances in one place.)

mkangia commented:
> The test was failing because a previous test was creating an app and not cleaning it up.
Oh cool... Good find..

> Dropped the include_remote=False param because remote apps get returned by the DB anyway and then skipped

I would still push to filter it already. Better to receive only what you need, as much as possible..

> so we might as well skip all non-Application instances in one place

Good point. I think you can just extract a method out of `get_apps_modules ` that does that so if its needed, one would not need write this again.
